src/Interpreters/TreeOptimizer.cpp

```diff
@@ -577,6 +713,15 @@ void TreeOptimizer::optimizeIf(ASTPtr & query, Aliases & aliases, bool if_chain_
         OptimizeIfChainsVisitor().visit(query);
 }
 
+void TreeOptimizer::optimizeAlterDeleteByPartitionKeyIfPossible(ASTPtr & query, const Context & context)
```

Comment:
I think it's a quite bad idea to perform this optimization on the syntax level. It would be much better to do it during query execution inside MergeTree table engine.

Reply:
This is a good idea worth to try. Let me think about it.